sdk/eventhub/azure-eventhub/azure/eventhub/_buffered_producer/_buffered_producer_dispatcher.py

```diff
@@ -0,0 +1,152 @@
+# --------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+import logging
+from threading import Lock
+from concurrent.futures import ThreadPoolExecutor
+from typing import Dict, Optional, List, Callable, TYPE_CHECKING
+
+from ._partition_resolver import PartitionResolver
+from ._buffered_producer import BufferedProducer
+from .._producer import EventHubProducer
+from ..exceptions import EventDataSendError, ConnectError
+
+if TYPE_CHECKING:
+    from .._producer_client import SendEventTypes
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class BufferedProducerDispatcher:
+    # pylint: disable=too-many-instance-attributes
+    def __init__(
+            self,
+            partitions: List[str],
+            on_success: Callable[["SendEventTypes", Optional[str]], None],
+            on_error: Callable[["SendEventTypes", Optional[str], Exception], None],
+            create_producer: Callable[..., EventHubProducer],
+            eventhub_name: str,
+            max_message_size_on_link: int,
+            *,
+            max_buffer_length: int = 1500,
+            max_wait_time: float = 1,
+            max_concurrent_sends: int = 1,
+            executor: Optional[ThreadPoolExecutor] = None,
+            max_worker: Optional[int] = None
+    ):
+        self._buffered_producers: Dict[str, BufferedProducer] = {}
+        self._partition_ids: List[str] = partitions
+        self._lock = Lock()
+        self._on_success = on_success
+        self._on_error = on_error
+        self._create_producer = create_producer
+        self._eventhub_name = eventhub_name
+        self._max_message_size_on_link = max_message_size_on_link
+        self._partition_resolver = PartitionResolver(self._partition_ids)
+        self._max_wait_time = max_wait_time
+        self._max_buffer_length = max_buffer_length
+        self._max_concurrent_sends = max_concurrent_sends
+        self._existing_executor = bool(executor)
+        self._executor = executor or ThreadPoolExecutor(max_worker)
```

Comment:
Wondering if we should use the number of partitions as the default worker count?

Reply:
that's a good point, we can but I'm thinking of the cases that partition count is kinda large e.g. 32/64 -- not sure how the perf would go when there're excessive amount of threads..

logged an issue for it

Reply:
https://github.com/Azure/azure-sdk-for-python/issues/24013